File: tests/test_decoder.py
import pytest

from riscv_sim import encoder as e
from riscv_sim.decoder import decode
from riscv_sim.errors import IllegalInstructionError


def test_decode_r_type_add():
    instr = decode(e.add(3, 1, 2))
    assert instr.fmt == "R"
    assert instr.mnemonic == "add"
    assert (instr.rd, instr.rs1, instr.rs2) == (3, 1, 2)


def test_decode_i_type_addi_positive_and_negative_immediate():
    pos = decode(e.addi(1, 0, 5))
    assert pos.mnemonic == "addi"
    assert (pos.rd, pos.rs1, pos.imm) == (1, 0, 5)

    neg = decode(e.addi(1, 1, -1))
    assert neg.imm == -1


def test_decode_shift_immediate_uses_shamt_field():
    instr = decode(e.slli(5, 1, 3))
    assert instr.mnemonic == "slli"
    assert instr.imm == 3


def test_decode_s_type_store():
    instr = decode(e.sw(2, 1, 0x40))
    assert instr.fmt == "S"
    assert instr.mnemonic == "sw"
    assert (instr.rs1, instr.rs2, instr.imm) == (2, 1, 0x40)


def test_decode_b_type_branch_negative_offset():
    instr = decode(e.bne(1, 0, -8))
    assert instr.fmt == "B"
    assert instr.mnemonic == "bne"
    assert instr.imm == -8


def test_decode_u_type_lui():
    instr = decode(e.lui(5, 0x12345000))
    assert instr.fmt == "U"
    assert instr.imm == 0x12345000


def test_decode_j_type_jal():
    instr = decode(e.jal(1, 0x800))
    assert instr.fmt == "J"
    assert instr.mnemonic == "jal"
    assert instr.imm == 0x800


def test_decode_system_instructions():
    assert decode(e.ecall()).mnemonic == "ecall"
    assert decode(e.ebreak()).mnemonic == "ebreak"


def test_decode_unknown_opcode_raises():
    with pytest.raises(IllegalInstructionError):
        decode(0b1111111)  # opcode bits that map to no known instruction


def test_instruction_str_formats_operands():
    assert str(decode(e.addi(1, 0, 5))) == "addi x1, x0, 5"
    assert str(decode(e.add(3, 1, 2))) == "add x3, x1, x2"

File: tests/test_decoder.cpp
#include "riscv_sim/decoder.hpp"

#include "riscv_sim/encoder.hpp"
#include "riscv_sim/errors.hpp"
#include "test_framework.hpp"

namespace e = riscv_sim::encoder;
using riscv_sim::decode;
using riscv_sim::IllegalInstructionError;

TEST_CASE("decode_r_type_add") {
    auto instr = decode(e::add(3, 1, 2));
    CHECK(instr.fmt == "R");
    CHECK(instr.mnemonic == "add");
    CHECK(instr.rd == 3);
    CHECK(instr.rs1 == 1);
    CHECK(instr.rs2 == 2);
}

TEST_CASE("decode_i_type_addi_positive_and_negative_immediate") {
    auto pos = decode(e::addi(1, 0, 5));
    CHECK(pos.mnemonic == "addi");
    CHECK(pos.rd == 1);
    CHECK(pos.rs1 == 0);
    CHECK(pos.imm == 5);

    auto neg = decode(e::addi(1, 1, -1));
    CHECK(neg.imm == -1);
}

TEST_CASE("decode_shift_immediate_uses_shamt_field") {
    auto instr = decode(e::slli(5, 1, 3));
    CHECK(instr.mnemonic == "slli");
    CHECK(instr.imm == 3);
}

TEST_CASE("decode_s_type_store") {
    auto instr = decode(e::sw(2, 1, 0x40));
    CHECK(instr.fmt == "S");
    CHECK(instr.mnemonic == "sw");
    CHECK(instr.rs1 == 2);
    CHECK(instr.rs2 == 1);
    CHECK(instr.imm == 0x40);
}

TEST_CASE("decode_b_type_branch_negative_offset") {
    auto instr = decode(e::bne(1, 0, -8));
    CHECK(instr.fmt == "B");
    CHECK(instr.mnemonic == "bne");
    CHECK(instr.imm == -8);
}

TEST_CASE("decode_u_type_lui") {
    auto instr = decode(e::lui(5, 0x12345000));
    CHECK(instr.fmt == "U");
    CHECK(instr.imm == 0x12345000);
}

TEST_CASE("decode_j_type_jal") {
    auto instr = decode(e::jal(1, 0x800));
    CHECK(instr.fmt == "J");
    CHECK(instr.mnemonic == "jal");
    CHECK(instr.imm == 0x800);
}

TEST_CASE("decode_system_instructions") {
    CHECK(decode(e::ecall()).mnemonic == "ecall");
    CHECK(decode(e::ebreak()).mnemonic == "ebreak");
}

TEST_CASE("decode_unknown_opcode_raises") {
    CHECK_THROWS_AS(decode(0b1111111), IllegalInstructionError);
}

TEST_CASE("instruction_str_formats_operands") {
    CHECK(decode(e::addi(1, 0, 5)).to_string() == "addi x1, x0, 5");
    CHECK(decode(e::add(3, 1, 2)).to_string() == "add x3, x1, x2");
}
